### PR TITLE
clearly indicate when users are external/pending on `/user` and `/user/<id>`

### LINKED ISSUE #412: visually distinguish external users on /user
Once we settle on a style in #410 lets use the same badge/label to identify clearly external users on /user

### PR DESCRIPTION
**Pull request by chris48s**
Closes #412
This has a bit of a merge conflict with #414 so lets work on getting that one merged first and I'll rebase this.

### CODE REVIEW

**Comment by chris48s** [comment]
I've rebase this to resolve the conflicts

File: ckanext/unhcr/templates/user/list.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  <tr>
+    <th>Username</th>
+    <th>Organisation</th>
+    <th>State</th>
+  </tr>
+  {% for user in page.items %}
+    <tr>
+      <td>{{ h.linked_user(user['name'], maxlength=20) }}</td>
+      <td>{% if h.user_is_external(user.User) %}external{% else %}UNHCR{% endif %}</td>
+      <td>{{ user.User.state }}</td>
```
Comment (amercader) [inline ckanext/unhcr/templates/user/list.html:14]:
```suggestion
      <td>{% if h.user_is_external(user.User) %}External{% else %}UNHCR{% endif %}</td>
      <td>{{ user.User.state.capitalize() }}</td>
```

File: ckanext/unhcr/templates/user/read_base.html [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        UNHCR
+      {% endif %}</dd>
+  </dl>
+  <dl>
+    <dt>{{ _('State') }}</dt>
+    <dd>
+      {% if user.state != 'active' %}
+        <span class="label label-warning">{{ user.state }}</span>
+      {% else %}
+        {{ user.state }}
```
Comment (amercader) [inline ckanext/unhcr/templates/user/read_base.html:44]:
```suggestion
        <span class="label label-warning">{{ user.state.capitalize() }}</span>
      {% else %}
        {{ user.state.capitalize() }}
```

File: ckanext/unhcr/templates/user/read_base.html [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  {% endif %}
+  <dl>
+    <dt>{{ _('Member Since') }}</dt>
+    <dd>{{ h.render_datetime(user.created) }}</dd>
+  </dl>
+  <dl>
+    <dt>{{ _('Organisation') }}</dt>
+    <dd>
+      {% if h.user_is_external(user) %}
+        <span class="label label-warning">external</span>
```
Comment (amercader) [inline ckanext/unhcr/templates/user/read_base.html:33]:
```suggestion
        <span class="label label-warning">External</span>
```

**Review by amercader — APPROVED** [review]
Looks good @chris48s , I just added some suggestions to capitalize the labels

**Comment by chris48s** [comment]
In https://github.com/okfn/ckanext-unhcr/pull/410 we went with lowercase using the same styles. Do you want me to update those to "External" (rather than "external") as well, or is there a reason for them to be different here?

**Comment by chris48s** [comment]
I realised there is another issue with this :(

When we call `user_show`, it only returns the email field if we are logged in as a sysadmin or the user we are showing. If I am some other user, I can view another user's profile but `user_show` doesn't return the external field, or the email. This causes us to throw `Invalid: 'not a user object'` in `user_is_external()`. In that case where we've got a user object but no `external` field and no email, I guess we need to fall back again and query the DB directly to bypass the auth and get the `external` value?

**Comment by amercader** [comment]
@chris48s the `external` property should always be set at the model level (ie `is_user_external` should call `self.email` and check the domain). Then we extend `user_show` to manually add `user['external'] = user_obj.external`. This way whenever we receive a user object, either a model object or a user dict we already have the `external` property available. We shouldn't need to call a `is_user_external` helper in the templates as this property will be in the user object already

Does this make sense?
So:

```python
# in plugin.py
def user_is_external(self):
    ''' 
    Returns True if user email is not in the managed internal domains.
    '''
    try:
        domain = self.email.split('@')[1]
    except AttributeError:
         # Internal sysadmin user does not have email
        if self.sysadmin:
            return False
        else:
            return True

    internal_domains = config.get('ckanext.unhcr.internal_domains', INTERNAL_DOMAINS),
        sep = ',')

    return domain not in internal_domains

class UnhcrPlugin(...):

    def update_config(self, config_):
        # ...
        User.external = property(user_is_external)


#in action.py

@toolkit.chained_action
def user_show(up_func, context, data_dict):
    user_dict = up_funct(context, data_dict)

    user_obj = context['user_obj']
    user_dict['external'] = user_obj.external
    return user_dict

```

**Comment by amercader** [comment]
Regarding capitalization, I believe it's better to Capitalize labels, but if you need to change it in many other places don't bother now.

**Comment by chris48s** [comment]
Yeah - I see where you're going with that, except we want the external flag from the user who we're showing not the user who is calling the action. i.e: we want to get .`external` for the `data_dict['id']` or `data_dict['name']` user because `context['user_obj']` is the calling user.
Any thoughts on how to deal with the case where we've called `user_list` with `return_query` and we're dealing with a ResultProxy instead of a dict or a User obejct? i.e: https://github.com/ckan/ckan/blob/7900935de6270ee80ec64a5378251dafedf3a0e4/ckan/logic/action/get.py#L894-L896
Anyways, I'll revisit this PR on Monday and have another go at it..

**Comment by chris48s** [comment]
Scratch all that in the last post - I'm getting confused about `context['auth_user_obj']` vs `context['user_obj']` and in the case where we've got a `ResultProxy`, we can access the `User` model object, so that's also fine.

This should be an improvement

**Review by amercader — APPROVED** [review]
This is good to go, moving the external property to the model makes everything clearer :+1:

**Pull request merged**